service/server.py

```diff
@@ -255,10 +255,33 @@ def get_address_lines(address_data):
         lines.append(address_data.get('postcode', None))
         lines.append(address_data.get('trail_info', None))
     non_empty_lines = [x for x in lines if x is not None]
+    # If the JSON doesn't contain the individual fields non_empty_lines will be empty
+    # Check if this is the case and if their is an address_string
+    if not non_empty_lines and address_data and address_data.get('address_string'):
+        non_empty_lines = format_address_string(address_data.get('address_string'))
     return non_empty_lines
 
 
-# This method attempts to retrieve the index polygon data for the entry
+def format_address_string(address_string):
+    # remove brackets and split the address string on commas
+    address_lines = re.sub('[\(\)]', '', address_string).split(', ')
+    # Strip leading and trailing whitespace and see if the last line is a just a postcode
+    last_line = address_lines[-1].strip()
+    if not re.search('^[A-Z]{1,2}[0-9R][0-9A-Z]? [0-9][A-Z]{2}$', last_line):
+        # If not, remove the line from address_lines, splt out the postcode and
+        # any preceeding text and trailing text and add them to address_lines
+        # as separate lines (if they exist)
+        del(address_lines[-1])
+        matches = re.match(r'(.*\b)\s?([A-Z]{1,2}[0-9R][0-9A-Z]? [0-9][A-Z]{2}\b)\s?(.*)?', last_line)
+        if matches.group(1) and len(matches.group(1).strip()) > 0:
```

Comment:
Python regexes can have named groups, which are great for readability. For example, instead of `matches.group(1)` we could refer to `matches.group('first_name')`, for example. Just to consider.